source/common/formatter/substitution_formatter.cc

```diff
@@ -54,5 +54,97 @@ const std::regex& SubstitutionFormatParser::commandWithArgsRegex() {
   // clang-format on
 }
 
+JsonFormatBuilder::FormatterEelements
+JsonFormatBuilder::fromStruct(const ProtobufWkt::Struct& struct_format) const {
+  sanitize_buffer_.clear();
+  raw_json_string_.clear();
+  output_.clear();
+
+  formatValueToFormatterEelements(struct_format);
+  if (!raw_json_string_.empty()) {
+    output_.push_back(JsonString{std::move(raw_json_string_)});
+  }
+  return std::move(output_);
+};
+
+void JsonFormatBuilder::formatValueToFormatterEelements(const ProtobufWkt::Value& value) const {
+  switch (value.kind_case()) {
+  case ProtobufWkt::Value::KIND_NOT_SET:
+  case ProtobufWkt::Value::kNullValue:
+    raw_json_string_.append(NullValue);
+    break;
+  case ProtobufWkt::Value::kNumberValue:
+    if (keep_value_type_) {
+      // Keep the number type in JSON.
+      raw_json_string_.append(fmt::to_string(value.number_value()));
```

Comment:
yes, let's focus on serializeDouble -- it is one place where we can also track improvements in the underlying C++ libraries for this,both for precision and accuracy.

Can you write a comment for this function that this is only called at startup?

I'm a little confused why we are calling a funciton that formats a number at startup.

Reply:
> yes, let's focus on serializeDouble -- it is one place where we can also track improvements in the underlying C++ libraries for this,both for precision and accuracy.

`serializeDouble` need take a Buffer::Instance as input parameter. I think maybe we also need to update the `serializeDouble`.

> Can you write a comment for this function that this is only called at startup?

Hmmm, rather than `at startup`, I think it's `at configuration loading (static or xDS)`. And this method is part of `JsonFormatBuiler` which is only used to help to create formatter instance. And the formatter instance will be used at runtime/core data paths.
So, maybe I should add a comment to `JsonFormatBuilder`.

> I'm a little confused why we are calling a funciton that formats a number at startup.

For example, if a raw double value is configured in the json format like following example:

```
json_format:
  raw_double_value: 12345
  format_string: %COMMAND%
```

According our current design (see more detail from https://github.com/envoyproxy/envoy/pull/35545#discussion_r1702048822), the double value should be serialized to json string piece directly when loading configuration.